cluster/store/meta_class.go

```diff
@@ -183,21 +183,39 @@ func mergeProps(old, new []*models.Property) []*models.Property {
 	return mergedProps
 }
 
-func (m *metaClass) AddTenants(nodeID string, req *command.AddTenantsRequest, v uint64) error {
+func (m *metaClass) AddTenants(nodeID string, req *command.AddTenantsRequest, replFactor int64, v uint64) error {
 	req.Tenants = removeNilTenants(req.Tenants)
 	m.Lock()
 	defer m.Unlock()
 
+	// TODO-RAFT: Optimize here and avoid iteration twice on the req.Tenants array
+	names := make([]string, len(req.Tenants))
+	for i, tenant := range req.Tenants {
+		names[i] = tenant.Name
+	}
+	// First determine the partition based on the node *present at the time of the log entry being created*
+	partitions, err := m.Sharding.GetPartitions(req.ClusterNodes, names, replFactor)
+	if err != nil {
+		return fmt.Errorf("get partitions: %w", err)
+	}
+
+	// Iterate over requested tenants and assign them, if found, a partition
 	for i, t := range req.Tenants {
 		if _, ok := m.Sharding.Physical[t.Name]; ok {
 			req.Tenants[i] = nil // already exists
 			continue
 		}
-
-		p := sharding.Physical{Name: t.Name, Status: t.Status, BelongsToNodes: t.Nodes}
+		// TODO-RAFT: Check in which cases can the partition not have assigned one to a tenant
+		part, ok := partitions[t.Name]
+		if !ok {
+			// TODO-RAFT: Do we want to silently continue here or raise an error ?
+			continue
```

Comment:
**suggestion:**  i'd say we can warn to avoid silent side effects

Reply:
I agree, the problem is that we don't have the logger available deep down here :s